internal/pipeline/types.go

```diff
@@ -12,20 +12,56 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
-// This file defines the type structures used by pipeline
+// Package pipeline provides struct definitions for Pipeline and utility
+// methods to read and write a pipeline resource.
 package pipeline
 
 import (
 	"sigs.k8s.io/kustomize/kyaml/yaml"
 )
 
+// Pipeline contains the content in pipeline.yaml file
+type Pipeline struct {
+	yaml.ResourceMeta `yaml:",inline"`
+	//  1. Sources to resolve as input to the pipeline. Possible values:
+	//  a) A relative path to a local package e.g. './base', '../foo'
+	//    The source package is resolved recursively.
+	//  b) Resources in this package using '.'. Meta resources such as the Kptfile, Pipeline, and function configs
+	//     are excluded.
+	//  c) Resources in this package AND all resolved subpackages using './*'
+	//
+	// Resultant list of resources are ordered:
+	// - According to the order of sources specified in this array.
+	// - When using './*': Subpackages are resolved in alphanumerical order before package resources.
+	//
+	// When omitted, defaults to './*'.
+	Sources []string `yaml:"sources,omitempty"`
+	// 2. Sequence of functions to run. Input of the first function is the resolved sources.
+	// Input of the second function is the output of the first function, and so on.
+	// Order of operations: generators, transformers, validators
+	//
+	// When omitted, defaults to NO-OP.
+	// 2.a  Sequence of KRM functions that generate resources.
+	Generators []Function `yaml:"generators,omitempty"`
+	// 2.b Sequence of KRM functions that transform resources.
+	Transformers []Function `yaml:"transformers,omitempty"`
+	// 2.c Sequence of KRM functions that validate resources.
+	// Validators are not permitted to mutate resources.
+	Validators []Function `yaml:"validators,omitempty"`
+}
+
 // Function defines an item in the pipeline function list
 type Function struct {
-	// Image is the image name that a function will use
+	// Image is the path of the function container image
+	// Image name can be a "built-in" function: kpt can be configured to use a image
+	// registry host-path that will be used to resolve the full image path in case
+	// the image path is missing (Defaults to gcr.io/kpt-functions-trusted).
+	// For example, the following resolves to gcr.io/kpt-functions-trusted/patch-strategic-merge.
+	//		image: patch-strategic-merge
 	Image string `yaml:"image,omitempty"`
 	// Config specifies an inline k8s resource used as the function config.
 	// Config, ConfigPath, and ConfigMap fields are mutually exclusive.
-	Config yaml.Node `yaml:"config,omitempty"`
+	Config FunctionConfig `yaml:"config,omitempty"`
```

Comment:
Why can't we use *yaml.Node directly ?

Reply:
I am assuming you are probably running in to some issue in making it a pointer.

If yaml.Node implement something for zero value, then we can skip making it pointer as we are doing it for configPath.

Reply:
The Node type in the latest go-yaml.v3 package does have a [`IsZero`](https://pkg.go.dev/gopkg.in/yaml.v3#Node.IsZero) method but it's not released so the version of the package that we are using in the kyaml doesn't have it. I have created a PR (https://github.com/kubernetes-sigs/kustomize/pull/3346) in kyaml to provide a helper function.

Reply:
In that case, I would suggest don't define FunctionConfig. Duplicate the IsZero code for now in kpt to use yaml.Node directly.